components/server/src/workspace/gitpod-server-impl.ts

```diff
@@ -1081,22 +1078,13 @@ export class GitpodServerImpl implements GitpodServerWithTracing, Disposable {
 
         const user = await this.checkUser("getWorkspaces");
 
-        const res = await this.workspaceDb.trace(ctx).find({
-            limit: 20,
-            ...options,
-            userId: user.id,
-            includeHeadless: false,
+        return this.workspaceService.getWorkspaces(user.id, options, async (workspace, instance) => {
+            if (instance) {
+                return this.guardAccess({ kind: "workspaceInstance", subject: instance, workspace: workspace }, "get");
+            } else {
+                return this.guardAccess({ kind: "workspace", subject: workspace }, "get");
+            }
         });
-        await Promise.all(res.map((ws) => this.guardAccess({ kind: "workspace", subject: ws.workspace }, "get")));
```

Comment:
Why did you not just leave the old guards as is? I think it would be nice if we can avoid passing the additional function to the new services and in this case it doesn't seem better as we are returning the workspace infos.

Reply:
Because in this case it felt like a trivial change. Yes, the "check vs outside" is sth that came later in the PR.

Reply:
But how is that trivial change better than now change? It pollutes the new workspaceService with unnecessary logic and contract. I would really like to keep that area healthy and keep the compromises in gitpod-server-impl.

Reply:
Somehow my comment from the workspace did not make it to GitHub (yet?): Originally I noted that this was not a proper filter at all, and intended to fixing this. But you are right, we don't need to. :+1: